libshortfin/src/shortfin/support/iree_helpers.h

```diff
@@ -105,41 +192,60 @@ class object_ptr {
     }
   }
   object_ptr(object_ptr &&other) : ptr(other.ptr) { other.ptr = nullptr; }
+  object_ptr &operator=(const object_ptr &other) = delete;
   object_ptr &operator=(object_ptr &&other) {
+    reset();
     ptr = other.ptr;
     other.ptr = nullptr;
     return *this;
   }
-  ~object_ptr() {
-    if (ptr) {
-      Helper::release(ptr);
-    }
-  }
+  ~object_ptr() { reset(); }
 
   // Constructs a new object_ptr by transferring ownership of a raw
   // pointer.
-  static object_ptr steal_reference(T *owned) { return object_ptr(owned); }
+  static object_ptr steal_reference(T *owned) {
+    Helper::steal(owned);
+    return object_ptr(owned);
+  }
+  // Constructs a new object_ptr by retaining a raw pointer.
   static object_ptr borrow_reference(T *owned) {
     Helper::retain(owned);
     return object_ptr(owned);
   }
   operator T *() const noexcept { return ptr; }
 
+  class Assignment {
+   public:
+    explicit Assignment(object_ptr *assign) : assign(assign) {}
+    ~Assignment() {
+      if (assign->ptr) {
+        Helper::steal(assign->ptr);
+      }
+    }
+
+    constexpr operator T **() noexcept {
+      return reinterpret_cast<T **>(&assign->ptr);
+    }
+
+   private:
+    object_ptr *assign = nullptr;
+  };
+
   // Releases any current reference held by this instance and returns a
   // pointer to the raw backing pointer. This is typically used for passing
   // to out parameters which are expected to store a new owned pointer directly.
-  T **for_output() {
+  constexpr Assignment for_output() noexcept {
     reset();
-    return &ptr;
+    return Assignment(this);
   }
 
   operator bool() const { return ptr != nullptr; }
   T *get() const { return ptr; }
-  void reset(T *other = nullptr) {
+  void reset() {
```

Comment:
This was a footgun since it allowed an unintended steal. Removed the capability as we have "steal_reference" and "borrow_reference" that do exactly what they say on the label.